source/extensions/common/matcher/domain_matcher.h

```diff
@@ -0,0 +1,244 @@
+
+#include "envoy/matcher/matcher.h"
+#include "envoy/network/filter.h"
+#include "envoy/server/factory_context.h"
+
+#include "source/common/matcher/matcher.h"
+
+#include "xds/type/matcher/v3/domain.pb.h"
+#include "xds/type/matcher/v3/domain.pb.validate.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Matcher {
+
+using ::Envoy::Matcher::DataInputFactoryCb;
+using ::Envoy::Matcher::DataInputGetResult;
+using ::Envoy::Matcher::DataInputPtr;
+using ::Envoy::Matcher::MatchResult;
+using ::Envoy::Matcher::MatchTree;
+using ::Envoy::Matcher::OnMatch;
+using ::Envoy::Matcher::OnMatchFactory;
+using ::Envoy::Matcher::OnMatchFactoryCb;
+using ::Envoy::Matcher::SkippedMatchCb;
+
+template <class DataType> struct DomainMatch {
+  size_t declaration_index_;
+  std::shared_ptr<OnMatch<DataType>> on_match_;
+
+  DomainMatch(size_t index, std::shared_ptr<OnMatch<DataType>> match)
+      : declaration_index_(index), on_match_(std::move(match)) {}
+};
+
+/**
+ * Domain matcher implementing ServerNameMatcher specification.
+ * Matches domains using exact lookups and wildcard patterns with proper precedence.
+ */
+template <class DataType> class DomainTrieMatcher : public MatchTree<DataType> {
+public:
+  DomainTrieMatcher(DataInputPtr<DataType>&& data_input,
+                    absl::optional<OnMatch<DataType>> on_no_match,
+                    absl::flat_hash_map<std::string, DomainMatch<DataType>> exact_matches,
+                    absl::flat_hash_map<std::string, DomainMatch<DataType>> wildcard_matches,
+                    std::shared_ptr<OnMatch<DataType>> global_wildcard_match)
+      : data_input_(std::move(data_input)), on_no_match_(std::move(on_no_match)),
+        exact_matches_(std::move(exact_matches)), wildcard_matches_(std::move(wildcard_matches)),
+        global_wildcard_match_(std::move(global_wildcard_match)) {
+    const auto input_type = data_input_->dataInputType();
+    if (input_type != Envoy::Matcher::DefaultMatchingDataType) {
+      throw EnvoyException(
+          absl::StrCat("Unsupported data input type: ", input_type,
+                       ", currently only string type is supported in domain matcher"));
+    }
+  }
+
+  MatchResult match(const DataType& data, SkippedMatchCb skipped_match_cb = nullptr) override {
+    const auto input = data_input_->get(data);
+    if (input.data_availability_ != DataInputGetResult::DataAvailability::AllDataAvailable) {
+      return MatchResult::insufficientData();
+    }
+
+    if (absl::holds_alternative<absl::monostate>(input.data_)) {
+      return MatchTree<DataType>::handleRecursionAndSkips(on_no_match_, data, skipped_match_cb);
+    }
+
+    const auto& domain = absl::get<std::string>(input.data_);
+    if (domain.empty()) {
+      return MatchTree<DataType>::handleRecursionAndSkips(on_no_match_, data, skipped_match_cb);
+    }
+
+    // Find best match using ServerNameMatcher specification algorithm
+    const auto best_match = findBestMatch(domain);
```

Comment:
I think this is not quite right, because we have to iterate over all possible match candidates and check the inner conditions as well.

PTAL as IPMatcher. It will go over candidates one-by-one and then decide based on the recursive matching (the exclusive flag is probably not necessary here).